File: NIREEKSHAK-main/apps/web/types/investigation.ts
export type RiskLevel = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';

export type InvestigationAction =
  | 'MARK_FOR_REVIEW'
  | 'UNDER_INVESTIGATION'
  | 'DISMISS_CLEAR'
  | 'ESCALATE_VIGILANCE';

export type TimelineStepStatus =
  | 'COMPLETED'
  | 'IN_PROGRESS'
  | 'DELAYED'
  | 'ANOMALOUS'
  | 'PENDING';

export type AnomalySignalType =
  | 'COST_ANOMALY'
  | 'DUPLICATE_WORK'
  | 'TIMELINE_VIOLATION'
  | 'GEOSPATIAL_ANOMALY'
  | 'AGENCY_CONCENTRATION';

export interface AnomalySignal {
  type: AnomalySignalType;
  title: string;
  severity: 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW';
  scoreImpact: number;
  description: string;
}

export interface ProjectHeader {
  projectId: string;
  title: string;
  sector: string;
  category: string;
  state: string;
  district: string;
  constituency: string;
  constituencyType: 'LOK_SABHA' | 'RAJYA_SABHA';
  mpName: string;
  mpHouse: string;
  sanctionDate: string;
  sanctionYear: string;
  sanctionOrderNumber: string;
  implementingAgency: string;
  nodalDepartment: string;
  sanctionedAmount: number; // in INR
  releasedAmount: number;   // in INR
  expenditureAmount: number; // in INR
  physicalProgressPercent: number;
  financialProgressPercent: number;
  currentStatus: 'UNDER_EXECUTION' | 'COMPLETED' | 'STALLED' | 'RECOMMENDED';
  lastUpdated: string;
}

export interface RiskScoreBreakdown {
  costDeviationScore: number;    // 0-100
  duplicateOverlapScore: number;  // 0-100
  timelineLatencyScore: number;   // 0-100
  agencyConcentrationScore: number; // 0-100
}

export interface RiskAssessment {
  overallScore: number; // 0-100
  riskLevel: RiskLevel;
  confidenceScore: number; // 0.00 - 1.00
  flaggedRulesCount: number;
  primaryDrivers: string[];
  breakdown: RiskScoreBreakdown;
  summaryText: string;
}

export interface CostBreakdownItem {
  component: string;
  projectCost: number;
  peerAverageCost: number;
  deviationPercent: number;
  anomalyFlag: boolean;
  notes: string;
}

export interface PeerDistributionPoint {
  percentile: string;
  cost: number; // in INR
  label: string;
}

export interface CostOutlierEvidence {
  thisProjectCost: number;
  unitMetric: string;
  unitValue: number;
  unitCost: number;
  peerUnitCostMedian: number;
  peerMedianCost: number;
  peerMeanCost: number;
  peerIqrLow: number;  // 25th percentile (Q1)
  peerIqrHigh: number; // 75th percentile (Q3)
  peerP95: number;     // 95th percentile
  peerMin: number;
  peerMax: number;
  deviationMultiplier: number; // e.g. 2.43x
  zScore: number;              // e.g. +3.42
  peerSampleSize: number;      // N
  baselineCategory: string;
  districtMedian: number;
  stateMedian: number;
  costBreakdown: CostBreakdownItem[];
  distributionCurve: PeerDistributionPoint[];
  statisticalObservations: string[];
}

export interface MatchedTokenDiff {
  text: string;
  isMatched: boolean;
  type?: 'identical' | 'near_match' | 'unique_current' | 'unique_matched';
}

export interface DuplicateMatchEvidence {
  matchScorePercent: number; // e.g. 91.4
  semanticMatchScore: number;
  syntacticMatchScore: number;
  matchedProjectId: string;
  matchedProjectTitle: string;
  matchedSanctionDate: string;
  matchedSanctionAmount: number;
  matchedImplementingAgency: string;
  matchedConstituency: string;
  matchedDistrict: string;
  matchedState: string;
  matchedLocationName: string;
  matchedGpsCoords: { lat: number; lng: number };
  currentGpsCoords: { lat: number; lng: number };
  distanceMeters: number;
  timeDeltaDays: number;
  isSameImplementingAgency: boolean;
  matchedPhrases: string[];
  currentDescriptionTokens: MatchedTokenDiff[];
  matchedDescriptionTokens: MatchedTokenDiff[];
  riskObservations: string[];
}

export interface TimelineAuditLog {
  date: string;
  event: string;
  recordedBy: string;
  severity: 'INFO' | 'WARNING' | 'CRITICAL';
}

export interface TimelineStep {
  stepId: string;
  stageKey: 'RECOMMENDATION' | 'SANCTION' | 'FUND_RELEASE' | 'EXECUTION' | 'COMPLETION';
  title: string;
  subtitle: string;
  status: TimelineStepStatus;
  plannedDate: string;
  actualDate: string | null;
  elapsedDays: number;
  standardSlaDays: number;
  deltaDays: number;
  disbursedPercentCumulative: number;
  physicalProgressPercentCumulative: number;
  isAnomalous: boolean;
  anomalyTitle?: string;
  anomalyDescription?: string;
  logs: TimelineAuditLog[];
}

export interface GeospatialEvidence {
  latitude: number;
  longitude: number;
  geoAccuracyMeters: number;
  geoSource: 'MOBILE_APP_ON_SITE' | 'PORTAL_MANUAL_ENTRY' | 'SURVEYOR_DPR';
  nearestPeerWorksCountWithin500m: number;
  clusterAnomalyDetected: boolean;
  satelliteClearanceScore: number; // 0-100
  cadastralLandId: string;
  landStatusNote: string;
}

export interface AgencyRiskEvidence {
  agencyName: string;
  contractorName: string;
  activeProjectsInBlockCount: number;
  totalWorksSanctionedCount: number;
  flaggedProjectsCount: number;
  allocationSharePercentInBlock: number;
  averageExecutionDelayDays: number;
  utilizationCertificatesPendingCount: number;
  riskRating: 'ELEVATED' | 'HIGH' | 'MODERATE' | 'LOW';
}

export interface AuditHistoryEntry {
  id: string;
  timestamp: string;
  officerId: string;
  officerName: string;
  officerDesignation: string;
  action: InvestigationAction;
  actionLabel: string;
  reviewNotes: string;
  tags: string[];
}

export interface ProjectInvestigationData {
  header: ProjectHeader;
  risk: RiskAssessment;
  signals?: AnomalySignal[];
  evidence: {
    costOutlier: CostOutlierEvidence;
    duplicateMatch: DuplicateMatchEvidence;
    geospatial: GeospatialEvidence;
    agencyRisk: AgencyRiskEvidence;
  };
  timeline: TimelineStep[];
  auditHistory: AuditHistoryEntry[];
  verificationStatus: {
    isReviewed: boolean;
    currentAction: InvestigationAction | null;
    reviewedBy?: string;
    reviewedAt?: string;
    reviewNotes?: string;
    tags?: string[];
  };
}

export interface VerifyInvestigationRequest {
  projectId: string;
  action: InvestigationAction;
  reviewNotes: string;
  officerId: string;
  officerName: string;
  officerDesignation: string;
  tags: string[];
  timestamp?: string;
}

export interface VerifyInvestigationResponse {
  success: boolean;
  message: string;
  updatedStatus: {
    projectId: string;
    action: InvestigationAction;
    reviewedBy: string;
    reviewedAt: string;
    auditLogId: string;
  };
}

File: NIREEKSHAK-datasets/apps/api/routes/investigation.py
from fastapi import APIRouter, HTTPException
import pandas as pd
from services.engine import engine

def round_floats(obj):
    if isinstance(obj, float):
        return round(obj, 4)
    elif isinstance(obj, dict):
        return {k: round_floats(v) for k, v in obj.items()}
    elif isinstance(obj, list):
        return [round_floats(x) for x in obj]
    return obj


router = APIRouter(prefix="/projects", tags=["Investigation & Evidence"])

@router.get("/{project_id}/investigation")
def get_investigation_data(project_id: str):
    if project_id in engine.df.index:
        row = engine.df.loc[project_id]
        if isinstance(row, pd.DataFrame):
            row = row.iloc[0]
    else:
        import hashlib
        hash_val = int(hashlib.md5(project_id.encode('utf-8')).hexdigest(), 16)
        idx = hash_val % len(engine.df)
        row = engine.df.iloc[idx]
        if isinstance(row, pd.DataFrame):
            row = row.iloc[0]
    
    # Parse reasons
    reasons = []
    try:
        raw_reasons = row.get("investigation_reasons", "[]")
        if isinstance(raw_reasons, str):
            reasons = eval(raw_reasons)
        elif isinstance(raw_reasons, list):
            reasons = raw_reasons
    except:
        pass
        
    signals = []
    for r in reasons:
        signals.append({
            "type": "COST_ANOMALY" if "cost" in r.lower() or "amount" in r.lower() else "DUPLICATE_WORK" if "duplicate" in r.lower() or "similar" in r.lower() else "TIMELINE_VIOLATION",
            "title": "Anomaly Detected",
            "severity": row.get("investigation_priority_v2_DERIVED", "HIGH"),
            "scoreImpact": 20,
            "description": str(r)
        })
        
    risk_score = int(row.get("investigation_score_v2_DERIVED", row.get("investigation_score", 0)))
    
    matched_id = row.get("nearest_work_id", "")
    duplicate_evidence = None
    if matched_id and matched_id in engine.df.index:
        peer = engine.df.loc[matched_id]
        duplicate_evidence = {
            "matchScorePercent": float(row.get("semantic_duplicate_score", 0.85)) * 100,
            "semanticMatchScore": float(row.get("semantic_duplicate_score", 0.85)) * 100,
            "syntacticMatchScore": float(row.get("semantic_duplicate_score", 0.85)) * 100,
            "matchedProjectId": matched_id,
            "matchedProjectTitle": peer.get("Work Description", ""),
            "matchedSanctionDate": str(peer.get("Sanction Date", "")),
            "matchedSanctionAmount": float(peer.get("Sanction Amount ( ₹ )", 0)),
            "matchedImplementingAgency": peer.get("Vendor Name", "Unknown"),
            "matchedConstituency": peer.get("Constituency", ""),
            "matchedDistrict": "",
            "matchedState": peer.get("State", ""),
            "matchedLocationName": "",
            "matchedGpsCoords": {"lat": 0, "lng": 0},
            "currentGpsCoords": {"lat": 0, "lng": 0},
            "distanceMeters": 0,
            "timeDeltaDays": 0,
            "isSameImplementingAgency": False,
            "matchedPhrases": [],
            "currentDescriptionTokens": [],
            "matchedDescriptionTokens": [],
            "riskObservations": []
        }
        
    cost_evidence = {
        "thisProjectCost": float(row.get("Sanction Amount ( ₹ )", 0)),
        "unitMetric": "project",
        "unitValue": 1,
        "unitCost": float(row.get("Sanction Amount ( ₹ )", 0)),
        "peerUnitCostMedian": float(row.get("peer_expected_amount", 0)),
        "peerMedianCost": float(row.get("peer_expected_amount", 0)),
        "peerMeanCost": float(row.get("peer_expected_amount", 0)),
        "peerIqrLow": float(row.get("peer_expected_lower", 0)),
        "peerIqrHigh": float(row.get("peer_expected_upper", 0)),
        "peerP95": float(row.get("peer_expected_upper", 0)) * 1.5,
        "peerMin": 0,
        "peerMax": float(row.get("peer_expected_upper", 0)) * 2,
        "deviationMultiplier": float(row.get("counterfactual_deviation", 1.0)),
        "zScore": float(row.get("robust_z", 0.0)),
        "peerSampleSize": 50,
        "baselineCategory": str(row.get("Work Category", "")),
        "districtMedian": 0,
        "stateMedian": 0,
        "costBreakdown": [],
        "distributionCurve": [],
        "statisticalObservations": []
    }

    return round_floats({
        "header": {
            "projectId": project_id,
            "title": row.get("Work Description", ""),
            "sector": row.get("Work Category", ""),
            "category": row.get("Work Category", ""),
            "state": row.get("State", ""),
            "district": row.get("Constituency", ""),
            "constituency": row.get("Constituency", ""),
            "constituencyType": "LOK_SABHA",
            "mpName": "Unknown",
            "mpHouse": "LOK_SABHA",
            "sanctionDate": str(row.get("Sanction Date", "")),
            "sanctionYear": "2023-2024",
            "sanctionOrderNumber": "UNKNOWN",
            "implementingAgency": row.get("Vendor Name", "Unknown"),
            "nodalDepartment": "Unknown",
            "sanctionedAmount": float(row.get("Sanction Amount ( ₹ )", 0)),
            "releasedAmount": float(row.get("Fund Disbursed Amount ( ₹ )", 0) if not pd.isna(row.get("Fund Disbursed Amount ( ₹ )")) else 0),
            "expenditureAmount": float(row.get("Amount Disbursed ( ₹ )", 0) if not pd.isna(row.get("Amount Disbursed ( ₹ )")) else 0),
            "physicalProgressPercent": 0,
            "financialProgressPercent": float(row.get("financial_fund_utilization_pct_DERIVED", 0)),
            "currentStatus": row.get("Payment Status", "UNDER_EXECUTION"),
            "lastUpdated": "2024-08-20",
        },
        "risk": {
            "overallScore": risk_score,
            "riskLevel": row.get("investigation_priority_v2_DERIVED", "LOW"),
            "confidenceScore": 0.85,
            "flaggedRulesCount": len(signals),
            "primaryDrivers": [str(r) for r in reasons],
            "breakdown": {
                "costDeviationScore": min(100, max(0, int(float(row.get("robust_z", 0)) * 10))),
                "duplicateOverlapScore": int(float(row.get("semantic_duplicate_score", 0)) * 100),
                "timelineLatencyScore": 0,
                "agencyConcentrationScore": 0,
            },
            "summaryText": str(row.get("investigation_recommendation", ""))
        },
        "signals": signals,
        "timeline": [],
        "evidence": {
            "costOutlier": cost_evidence,
            "duplicateMatch": duplicate_evidence,
            "geospatial": {
                "latitude": 27.5,
                "longitude": 79.5,
                "geoAccuracyMeters": 15,
                "geoSource": "PORTAL_MANUAL_ENTRY",
                "nearestPeerWorksCountWithin500m": 0,
                "clusterAnomalyDetected": False,
                "satelliteClearanceScore": 0,
                "cadastralLandId": "UNKNOWN",
                "landStatusNote": "N/A"
            },
            "agencyRisk": {
                "agencyName": row.get("Vendor Name", "Unknown"),
                "totalProjectsActive": 0,
                "totalProjectValue": 0,
                "anomalousProjectsCount": 0,
                "utilizationCertificatesPendingCount": 0,
                "riskRating": "LOW"
            }
        },
        "auditHistory": [],
        "verificationStatus": {
            "isReviewed": False,
            "currentAction": "FLAGGED",
            "reviewedBy": "",
            "reviewedAt": ""
        }
    })
from schemas import SimilarProjectsResponse, TimelineResponse, EvidenceResponse

@router.get("/{project_id}/similar", response_model=SimilarProjectsResponse)
def get_similar_projects(project_id: str):
    similar = engine.get_similar_projects(project_id)
    return {"items": similar}

@router.get("/{project_id}/timeline", response_model=TimelineResponse)
def get_timeline(project_id: str):
    return {
        "events": [
            {
                "type": "RECOMMENDATION",
                "date": "2024-03-04",
                "status": "completed"
            }
        ]
    }

@router.get("/{project_id}/evidence", response_model=EvidenceResponse)
def get_evidence(project_id: str):
    ev = engine.get_evidence(project_id)
    if not ev:
        raise HTTPException(status_code=404, detail="Evidence not found")
    return ev
